add polymorthism to gesture services

josxha opened the pull request.

josxha committed:
introduce SetTapDownDetailsMixin

josxha committed:
add ProgressableGestureService , BaseLongPressGestureService, remove SetTapDownDetailsMixin

JaffaKetchup commented on lib/src/map/gestures/services/key_trigger_drag_rotate.dart:
```diff
@@ -1,8 +1,11 @@
-class KeyTriggerDragRotateGestureService extends BaseGestureService {
+/// is by default a CTRL +.
+///
+/// Can't extend from [_ProgressableGestureService] because of different
+/// method signatures.
```
To avoid this, can we not use generic typing on the `_ProgressableGestureService`, and have the `start` argument be generically typed, and potentially named?
Then normally it can be `ScaleUpdateStart`, but in this case it can be `Null`? So the restriction on the generic type would be `ScaleUpdateStart?`?

josxha replied:
While I agree that it makes sense to have generics for variants of classes (e.g. Point<double> to have a double variant of Point), I don't think that introducing a generic for the type of a parameter type for only one function is good practice. Why try to force an interhitance on them if they **look** similar but **aren't** actually that similar?

---

Edit: Same for `TertiaryTapGestureService.submit()` the idea of the service methods was to match the signature of the callbacks onTeniaryTap would have had a `TapUpDetails` parameter. Because we force it to be a `SingleShotGestureService`, there needs to be additional code in the `MapInteractiveViewer`:
```dart
        onTertiaryTapDown: _tertiaryTap?.setDetails,
        onTertiaryTapCancel: _tertiaryTap?.reset,
        onTertiaryTapUp:
            _tertiaryTap == null ? null : (_) => _tertiaryTap?.submit(),
```

JaffaKetchup replied:
Ok, I understand that.

---

I see what you mean, but I would argue that `void submit(TapUpDetails _)` isn't ideal either. I'm struggling to understand why teritary tap is different to secondary tap or single tap? Why does it have a different signature? Am I just being thick?

josxha replied:
@JaffaKetchup The other gestures use the normal tap callback. Only the `onTertiaryTap` isn't provided by the `GestureDetector`.
We could use the `-up` callbacks for the tap, secondary-tap and tertiary-tap gestures. 

However the double-tap gesture does **not** have a `-up` callback. That's why we have to use the `onDoubleTapDown` callback.

My logic while implementing was to prefer using the normal tap callback and to use the -up callback if the other is not not available. if you want I can change this to whatever you like.

josxha replied:
discussed internally

JaffaKetchup reviewed (commented):
LGTM, except the one thing below. Is there no good way to solve this? I'm not sure if what I propose will actually work very nicely, it might just result in a less friendly API. It's up to you, happy enough to ignore this and approve if you disagree.

josxha committed:
Merge branch 'gesture-handling' into gesture-service-setDetails-as-mixin

josxha committed:
clean up methods

no longer necessary to cancel animations within the services, any pointer-down event stop ongoing gesture animation now.

josxha committed:
change abtract classes that have no method bodies or fields to interfaces

josxha commented:
I merged the upstream pull request. The services no longer stop animations because animations get stopped now before a gesture wins.
This changes `_BaseLongPressGestureService` and `_ProgressableGestureService` to have no implemented method bodies.
